### PR TITLE
google-aec: Component unit test

### PR DESCRIPTION
**Pull request by lkoenig**
This unit tests ensure all function got called. It does not verify audio
processing behaviour but allow testing for memory allocation/free and
hence tests leakage.

Signed-off-by: Lionel Koenig <lionelk@google.com>

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@singalsu any comment ?

File: test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+#include <setjmp.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <cmocka.h>
+#include <sof/audio/component.h>
+#include <ipc/topology.h>
+#include <sof/audio/component_ext.h>
+#include <sof/lib/uuid.h>
+
+#define FRAME_SIZE 48
```
Comment (cujomalainey) [inline test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:18]: i thought we do 10ms?

File: test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	google_rtc_audio_processing_state->ref->stream.rate = SAMPLERATE;
+	google_rtc_audio_processing_state->out->stream.rate = SAMPLERATE;
+	google_rtc_audio_processing_state->mic->source->ipc_config.type =
+		SOF_COMP_DAI;
+	google_rtc_audio_processing_state->ref->source->ipc_config.type =
+		SOF_COMP_DEMUX;
+
+	comp_trigger(google_rtc_audio_processing_state->dev,
+		     COMP_TRIGGER_PREPARE);
+	comp_prepare(google_rtc_audio_processing_state->dev);
```
Comment (cujomalainey) [inline test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:85]: do we need to check these return codes?

**Comment by lkoenig** [comment]
We do 10ms. It should work with almost any size. The processing will only occure for every 10ms of audio.
The component should not have to care how much the period size is.

**Comment by lkoenig** [comment]
I added checks for the return codes.
I also added some supported use-case in term of number of channels and period size.

This test won't work on main due to the UUID issue I mentioned earlier: in `src/include/sof/lib/uuid.h` the macros are defined to return `NULL`when `CONFIG_LIB` is defined. Which will cause an issue in the test setup when trying to construct the component. @singalsu I was wondering if you had an idea here or if I am doing something wrong ?

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by cujomalainey** [comment]
@lkoenig that sounds like something I am trying to undo. I need UUIDs in CONFIG_LIB for the fuzzer.

**Comment by singalsu** [comment]
@lkoenig I was OOO for two weeks, have you been able to solve this? I haven't done cmocka test cases for components after UUID introduction so I may have well missed this.

**Comment by lkoenig** [comment]
@cujomalainey let me know when you found a solution so I cna update this PR.
@singalsu I haven't solve it yet (counting on @cujomalainey) to figure something smart up.

**Review by lgirdwood — APPROVED** [review]
I think the UT looks fine, @singalsu any comments ?

**Comment by lgirdwood** [comment]
> It depends on how we mock the build if we add back the linker script or not to do dynamic registration. @lgirdwood do you know the plan here?

@singalsu has some work to look at the changes needed to run actual processing tests as part of CI using testbench, this would include the linkage. We have made some Testbench cleanup and improvement, but not sure if everything is in place yet to run AEC testing ?

File: test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	comp_update_buffer_produce(google_rtc_audio_processing_state->mic,
+				   length_bytes);
+	comp_update_buffer_produce(google_rtc_audio_processing_state->ref,
+				   length_bytes);
+}
+
+static void verify_s16_to_s16(struct comp_dev *dev, struct comp_buffer *mic,
+			      struct comp_buffer *ref, struct comp_buffer *out)
+{
```
Comment (singalsu) [inline test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:172]: The magic is missing :)

**Review by singalsu — COMMENTED** [review]
Looks very promising, thanks!

**Comment by singalsu** [comment]
> @singalsu has some work to look at the changes needed to run actual processing tests as part of CI using testbench, this would include the linkage. We have made some Testbench cleanup and improvement, but not sure if everything is in place yet to run AEC testing ?

I have been able to build and run the mockup version of this AEC in testbench. The test topologies for such purpose needs some thinking but I could make a quick draft PR to show how and how to get CI to test it. I have also work done to run all current process components in testbench in CI but it needs polishing and a new PR. It was earlier part of other PR and not really in that PR context so it was postponed.

**Comment by lkoenig** [comment]
Sorry for the long delay, I was busy on other projects.

File: test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c (hunk 3 above)
Comment (cujomalainey) [reply test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:199]: @lkoenig just want to confirm this is still not running any actual data validation

**Comment by cujomalainey** [comment]
Looks LGTM to me, just curious if we will do any data validation, if not then I will mark as approved

**Comment by lkoenig** [comment]
One think we can do is check that we have audio energy on the output when energy is fed on the input. But event that is not really reliable as the AEC might not be that transparent. That would work for the mock though. If you feel this is something worth pursuing I can add a very small check for that.

**Comment by lkoenig** [comment]
One think we can do is check that we have audio energy on the output when energy is fed on the input. But event that is not really reliable as the AEC might not be that transparent. That would work for the mock though. If you feel this is something worth pursuing I can add a very small check for that.

**Comment by cujomalainey** [comment]
Lets sync on this, I just want to make sure this gets the best value for its effort.

**Comment by kv2019i** [comment]
Converting to draft, no activity in 3 months.

**Comment by cujomalainey** [comment]
I think this can be closed out until we get further updates

**Pull request closed without merging**